### ISSUE TITLE
[Performance] Why is ONNX Runtime GPT2 Model Significantly Slower than PyTorch

### ISSUE THREAD

**Issue by dyerdave-gh** [microsoft/onnxruntime#13105]
### Describe the issue

I was intrigued by @tianleiwu 's [excellent blog post](https://medium.com/microsoftazure/accelerate-your-nlp-pipelines-using-hugging-face-transformers-and-onnx-runtime-2443578f4333) about speeding up inference with ONNX Runtime (specifically with GPT2), however, have only seen slower results during inference with real data, trying a variety of methods.  I feel like I'm missing something, and was wondering if you could point me in the right direction.  I see speed increases with the following dummy code:

```
from transformers import AutoTokenizer, AutoModelForCausalLM
import onnxruntime as ort
import torch
from onnxruntime import ExecutionMode, InferenceSession, SessionOptions

DEVICE = 'cuda'
model = AutoModelForCausalLM.from_pretrained("gpt2").to(DEVICE)
dummy_tokenizer = AutoTokenizer.from_pretrained("gpt2", use_fast=True)
dummy_tokenizer.pad_token = dummy_tokenizer.eos_token

dummy_inputs = dummy_tokenizer("My dog is a", return_tensors="pt")
dummy_input_ids = dummy_inputs['input_ids'].to(DEVICE)
dummy_att_mask = dummy_inputs["attention_mask"].to(DEVICE)
dummy_args = (dummy_input_ids, None, dummy_att_mask)

dummy_ortvalue_input_ids = ort.OrtValue.ortvalue_from_numpy(dummy_input_ids.cpu().numpy(), 'cuda', 0)
dummy_ortvalue_att_mask = ort.OrtValue.ortvalue_from_numpy(dummy_att_mask.cpu().numpy(), 'cuda', 0)
dummy_ortvalue_args = (dummy_ortvalue_input_ids, None, dummy_ortvalue_att_mask)

torch.onnx.export(model, dummy_args, "models/AutoModelForCausalLM.onnx", input_names = ['ort_input_ids', 'ort_att_mask'])

options = SessionOptions()
options.intra_op_num_threads = 1
options.execution_mode = ExecutionMode.ORT_SEQUENTIAL
session = InferenceSession('models/AutoModelForCausalLM.onnx', options, providers=['CUDAExecutionProvider'])
```
![image](https://user-images.githubusercontent.com/93223986/192357775-03754ff3-c4ce-47a2-924f-23b760098819.png)

However, when I use real data, I see significantly slower results -- with ONNX often being in the ~40% slower range.  
I've tried OLive, ONNX, and ONNX Runtime Auto-optimizers, I've tried constant folding, many different `SessionOptions` (including turning `intra_op_num_threads` up, and `ORT_SEQUENTIAL`) with no luck.  I've tried iobinding, and basically anything I could find in the [performance tuning guide](https://onnxruntime.ai/docs/performance/tune-performance.html), but I always get a slower inference result with ONNX.

![image](https://user-images.githubusercontent.com/93223986/192359204-3c26b80f-8239-4340-a5be-d67338277a31.png)

The only two things that have sped up my inference project is turning up the `intra_op_num_threads` to > 1, and making sure to `run_with_ortvalue` with ortvalues that are bound to CUDA.  Both of those together resulted in about a 30% speed increase, but still nowhere near PyTorch.  However, to be fair, I'm not entirely sure I'm using all the ORT tools at my disposal.  I would love an update to the blog post showing current speed increases and updated example notebooks!

Is this an optimization problem, or am I just missing something about the limitations of the ONNX Runtime project? The original blog / guide from 2020 has a number of dead links, so I'm not sure how I can achieve even moderately close to the same results that you're receiving in the article.  Thanks!

### To reproduce

I can't share my code and model here on a public forum, but you can see the same issue if you run the [ONNX PyTorch CV Examples using the Microsoft Learn Tutorial and the FashionMNIST Data Set](https://onnxruntime.ai/docs/get-started/with-python.html#quickstart-examples-for-pytorch-tensorflow-and-scikit-learn), where we see similar results.  I even switched to a P3 (I'm currently running on a G5XL) because I wondered if the older GPUs would run faster (they didn't)
<img width="633" alt="image" src="https://user-images.githubusercontent.com/93223986/192363297-92eeef0a-f83a-4baf-a804-c6323bce8039.png">

### Urgency

I probably waited too long to put in this issue, but I wanted to make sure I tried everything I could.  As it stands now, I am able to give ONNX about a week  of my attention and if I can't speed up inference, then it will be time to move on.

### Platform

Linux

### OS Version

5.10.102-99.473.amzn2.x86_64 (ID_LIKE="centos rhel fedora")

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.11.10

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA: 11.1, CuDNN: 8.0.5, NVCC: cuda_11.1.TC455_06.29190527_0, 

### Is this a quantized model?

No

**Comment by tianleiwu** [comment]
@dyerdave-gh, current optimization is target for GPT-2 model with past state (past_key_value). If you uses past state, you will need use it with IO Binding, and try convert model to mixed precision (and keep past state in float16) to improve performance without significant loss in accuracy.

You might try out BeamSearch or GreedySearch integration with GPT-2:
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/convert_generation.py
It is significantly (2X ~ 4X) faster than PyTorch in some of our test settings.

**Comment by dyerdave-gh** [comment]
Hi @tianleiwu thank you so much for the assistance. We are seeing similar, faster, results to what you explained above, using IO Binding.  However, I noticed that the benchmark doesn't include converting the `ort_results` (of dtype ortvalue) to a numpy array.  We need faster inference, including the final step, which actually takes about 60% of the total time.  Is there any faster way than `.numpy()` to get the data out of ortvalue dtypes to something we can decode and use?

**Comment by tianleiwu** [comment]
@dyerdave-gh, 
OrtValue is a block of memory. If you use C++, you can directly read the values. 

If the time takes 60%, is it possible to move some processing (like beam search etc) to ORT, so that only a small tensor (like word IDs)  is output.

For python API, you can bind the output to PyTorch tensor, then use it for post-processing (like convert word ID to string etc). See the last example of IO Binding in https://onnxruntime.ai/docs/api/python/api_summary.html.
